### PR TITLE
ENH: add IntervalIndex.get_loc_exact to look for exact matches only

### LINKED ISSUE #19349: BUG/inconsistency: IntervalIndex.get_loc gives a location for non-interval inputs
This one is a bit complex to explain, but I'll do my best.

Currently ``IntervalIndex.get_indexer`` fails if the other index doesn't contain  ``Interval`` only (also another bug, but let's keep it simple here).

The underlying issue is that ``IntervalIndex.get_indexer`` depends on ``IntervalIndex.get_loc`` which is ambigous for how it treats number inputs:

```python
>> ii = pd.IntervalIndex.from_breaks([0,1,2,3])
>> ii.get_loc(pd.Interval(1, 2))
1  # ok
>> ii.get_loc(1)  # do we mean exactly 1, or if an interval contains the number 1?
1  # ambigous
```
The issue is that ``get_loc`` returns the location for both exact matches and inexact matches (i.e. if the number input is in an interval). For the purposes of ``get_indexer`` however, this behavious fails, as ``get_indexer`` needs to find exact matches only.

See https://github.com/pandas-dev/pandas/pull/19021#issuecomment-359544778 for further discussion.

## Solution
A solution could be adding a boolean  ``strict`` parameter to ``IntervalIndex.get_loc`` or add a new method ``get_loc_exact_only``.

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #19349
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR add a 'exact' option to the parameter ``method`` on ``IntervalIndex.get_loc`` for accepting exact matches only.

In addition to it's direct advantage, it will later make it possible to make a PR to allow ``IntervalIndex.get_indexer`` calls with the other index having non-Interval values, and hence allow mixing mix Interval with non-intervals in ``pd.Index(..., dtype=object)`` instances. This would fix the remaining issue with #19021.

### CODE REVIEW

**Comment by jschendel** [comment]
There's a private method that looks like it's supposed to do this:

https://github.com/pandas-dev/pandas/blob/9872d6757e5117dce070981141cee562f675694e/pandas/core/indexes/interval.py#L870-L880

It doesn't appear to be working at all though:
```python
In [2]: ii = pd.interval_range(0, 4)

In [3]: ii.get_loc(pd.Interval(0, 1))
Out[3]: 0

In [4]: ii._get_loc_only_exact_matches(pd.Interval(0, 1))
---------------------------------------------------------------------------
KeyError: Interval(0, 1, closed='right')
```
The issue appears to be that `Index(self._multiindex.values)` creates an index of tuples, not intervals, so passing an `Interval` object will always raise a `KeyError`.  Likewise, due to the first line, passing anything other than an `Interval` object will also raise a `KeyError`.
```python
In [5]: pd.Index(ii._multiindex.values)
Out[5]: Index([(0, 1), (1, 2), (2, 3), (3, 4)], dtype='object')
```
Might be best to implement this in `_get_loc_only_exact_matches` instead?

File: pandas/tests/indexes/interval/test_interval.py [hunk 1]
```diff
@@ -739,6 +743,10 @@ def test_get_loc_interval(self):
         pytest.raises(KeyError, self.index.get_loc, Interval(2, 3))
         pytest.raises(KeyError, self.index.get_loc,
                       Interval(-1, 0, 'left'))
+        # GH19349
+        assert self.index.get_loc(Interval(0, 1), method='exact') == 0
+        with pytest.raises(KeyError):
+            self.index.get_loc(Interval(0, 0.5), method='exact')
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval.py]: Might be beneficial to also test than an `Interval` that only differs by `closed` returns a `KeyError` as well, e.g. `Interval(0, 1, closed='neither')`.

**Comment by jreback** [comment]
yes, rather than adding kwargs to ``.get_loc`` much better to fix the private method here.

**Comment by topper-123** [comment]
I can look into the private method.

@jreback , do you mean you are negative wrt. ``method='exact'`` in the public interface? I think it will be useful to have in many cases, e.g.

```python
>>> ii = pd.interva_range(0, 3)
>>> ii.get_loc(pd.Interval(1,3))
slice(1, 3, None)
>>> ii.get_loc(pd.Interval(1,3), method='exact')
KeyError: ...
```

Above, we've answered two different questions, both reasonable to ask, IMO.

**Comment by topper-123** [comment]
In ``_get_loc_only_exact_matches`` I've converted the line ``return Index(self._multiindex.values).get_loc(key)`` to ``return super(IntervalIndex, self)._engine.get_loc(key)``. Another option would be to do ``self.astype(object).get_loc(key)``, which is more readable, but slower, presumedly because of the need to cast.

The super method is faster, around 13 ms, while the astype method is around 75 ms, so I've chosen that.

Also, ``_get_loc_only_exact_matches`` currently only works for unique indices, while ``get_loc`` should work with non-unique indices.

**Comment by jreback** [comment]
@topper-123 we are not going to change the ``.get_loc`` api. it is very simple and needs to stay that way. As I said I am not above exposing the private function to make an exact match.

**Comment by topper-123** [comment]
@jreback , adding this *keeps* things simple IMO: If you look at the doc string for ``pd.Index`.get_loc``, the ``method`` is:
```
method : {None, 'pad'/'ffill', 'backfill'/'bfill', 'nearest'}, optional
    * default: exact matches only.
    * ...
```

It's logical that ``IntervalIndex.get_loc`` should be able to revert back to exact matches only, which is the default for all other index types.

**Comment by topper-123** [comment]
I've added a ``get_loc_exact`` method instead of using ``method='exact'`` in ``get_loc``.

**Comment by topper-123** [comment]
Nice catch.

I've changed the call to ``return self.astype(object).get_loc(key)`` and added a test for the condition you pointed out.

A shame, as the super method would have been faster if your issue would not have surfaced.

File: pandas/core/indexes/interval.py [hunk 2]
```diff
@@ -1003,6 +1004,47 @@ def get_loc(self, key, method=None):
...
+            The label we want to find locations for. Must have type
+            :class:`Interval`
+        method : {None}, optional
+            * default: matches where the label exactly matches a given
+              :class:`Interval`.
+
+        Returns
+        -------
+        loc : int if unique index, slice if monotonic index, else mask
+
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: add Raises section

File: pandas/core/indexes/interval.py [hunk 3]
```diff
@@ -1003,6 +1004,47 @@ def get_loc(self, key, method=None):
             else:
                 return self._engine.get_loc(key)
 
+    def get_loc_exact(self, key, method=None):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: add the same method in Index, dispatch to .get_loc

File: pandas/core/indexes/interval.py [hunk 4]
```diff
@@ -1003,6 +1004,47 @@ def get_loc(self, key, method=None):
...
+
+        >>> index = pd.IntervalIndex([i1, i2, i1])
+        >>> index.get_loc_exact(i1)
+        array([ True, False,  True], dtype=bool)
+
+        See Also
+        --------
+        get_loc
+        """
+        return self.astype(object).get_loc(key, method=method)
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: see above

File: pandas/tests/indexes/interval/test_interval.py [hunk 5]
```diff
@@ -497,6 +497,23 @@ def test_get_loc_interval(self):
...
+            self.index.get_loc_exact(1)
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(0, 1, 'left'))
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(0, 0.5))
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(2, 3))
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(-1, 0, 'left'))
+        # See #19353#issuecomment-364295029
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: what is this last test?

File: pandas/tests/indexes/interval/test_interval.py [hunk 6]
```diff
@@ -497,6 +497,23 @@ def test_get_loc_interval(self):
         pytest.raises(KeyError, self.index.get_loc,
                       Interval(-1, 0, 'left'))
 
+    # To be removed, replaced by test_interval_new.py (see #16316, #16386)
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: why are you adding this comment?

File: doc/source/whatsnew/v0.23.0.txt [hunk 7]
```diff
@@ -507,6 +507,7 @@ Other API Changes
 - Set operations (union, difference...) on :class:`IntervalIndex` with incompatible index types will now raise a ``TypeError`` rather than a ``ValueError`` (:issue:`19329`)
 - :class:`DateOffset` objects render more simply, e.g. "<DateOffset: days=1>" instead of "<DateOffset: kwds={'days': 1}>" (:issue:`19403`)
 - :func:`pandas.merge` provides a more informative error message when trying to merge on timezone-aware and timezone-naive columns (:issue:`15800`)
+- New :meth:`IntervalIndex.get_loc_exact` has been added to find exact Interval matches only (:issue:`19349`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: move to Other Enhancements

Comment (jreback) [reply doc/source/whatsnew/v0.23.0.txt]: add to the api.rst

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (jreback) [inline pandas/core/indexes/interval.py]: this seems quite inefficient. a) can you add an asv, b) I would do this in the cython code itself that implements get_loc (this can be a further refinement / PR).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (topper-123) [reply pandas/core/indexes/interval.py]: The super method was faster, around 13 ms, while this is around 75 ms, so it's slower.

I don't have experience with cython, so can't take on those changes in get_loc, sorry.

File: pandas/tests/indexes/interval/test_interval.py (hunk 5 above)
Comment (topper-123) [reply pandas/tests/indexes/interval/test_interval.py]: this tests if the caching issue showed by @jschendel exists.

Comment (jreback) [reply pandas/tests/indexes/interval/test_interval.py]: ok then write that as a comment. an issue reference is good, but a comment + reference even better. add blank lines in between sections of tests.

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (jreback) [reply pandas/core/indexes/interval.py]: ok that's fine an asv for now is prob best.

Comment (topper-123) [reply pandas/core/indexes/interval.py]: I've looked into asv and the speed issue. 

```python
>>> idx = IntervalIndex.from_breaks(np.arange(1000001))
>>> i = idx[80000]
>>> %t idx.get_loc(i)
25.2 µs ± 1.31 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
>>> %t idx.get_loc_exact(i)
1.33 s ± 46.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I doubt this acceptable?

Comment (jreback) [reply pandas/core/indexes/interval.py]: wait, so this is *much* slower?

Comment (topper-123) [reply pandas/core/indexes/interval.py]: Yes. I think this speed issue is beyound me, so I'm stuck. Maybe it's something wrong with Interval type conversion?

Comment (jreback) [reply pandas/core/indexes/interval.py]: so why don't you just do .get_loc and then check if its exact after?

Comment (topper-123) [reply pandas/core/indexes/interval.py]: Ok, that works very nicely:  41.5 µs for ``get_loc_exact`` vs 33.1 µs for ``get_loc``.

so there's no speed issue now at all.

File: pandas/core/indexes/interval.py [hunk 8]
```diff
@@ -1003,6 +1004,63 @@ def get_loc(self, key, method=None):
...
+        get_loc
+        """
+        all_matches = self.get_loc(key, method=method)
+        if is_scalar(all_matches):
+            if self[all_matches] == key:
+                return all_matches
+            raise KeyError(key)
+        else:
+            exact_matches = self[all_matches] == key
+            if (exact_matches).all():
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: you can move this up to avoid duplication

File: pandas/core/indexes/interval.py [hunk 9]
```diff
@@ -1003,6 +1004,63 @@ def get_loc(self, key, method=None):
...
+                return all_matches
+            elif (exact_matches).any():
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: this can also be simplified i think

File: pandas/tests/indexes/interval/test_interval.py [hunk 10]
```diff
@@ -497,6 +497,24 @@ def test_get_loc_interval(self):
...
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(0, 0.5))
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(2, 3))
+        with pytest.raises(KeyError):
+            self.index.get_loc_exact(Interval(-1, 0, 'left'))
+
+        # The below tests if get_loc_exact interferes with caching
+        # used for index.get_loc. See #19353#issuecomment-364295029
+        self.index.get_loc(Interval(0, 1))
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: assert that this is something

File: pandas/tests/indexing/interval/test_interval.py [hunk 11]
```diff
@@ -136,7 +136,7 @@ def test_with_slices(self):
             s[Interval(3, 6):]
 
         expected = s.iloc[3:5]
-        result = s[[Interval(3, 6)]]
+        result = s[Interval(3, 6)]
```
Comment (jreback) [inline pandas/tests/indexing/interval/test_interval.py]: why is this changed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/interval.py (hunk 9 above)
Comment (topper-123) [reply pandas/core/indexes/interval.py]: I don't think so: in the ``exact_matches.all()`` case, all_matches might be a slice object, so ``all_matches[exact_matches]`` would then fail.

File: pandas/tests/indexing/interval/test_interval.py (hunk 11 above)
Comment (topper-123) [reply pandas/tests/indexing/interval/test_interval.py]: This is unneeded, I've changed it back.

**Comment by topper-123** [comment]
All concerns have been addessed, AFAICS.

File: pandas/core/indexes/interval.py [hunk 12]
```diff
@@ -1003,6 +1003,63 @@ def get_loc(self, key, method=None):
...
+        See Also
+        --------
+        get_loc
+        """
+        all_matches = self.get_loc(key, method=method)
+        exact_matches = self[all_matches] == key
+        if is_scalar(all_matches):
+            if exact_matches:
+                return all_matches
+            raise KeyError(key)
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: this can be simplified

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
green.

**Comment by jorisvandenbossche** [comment]
Is it really needed to add a method to all Index objects only for use in `IntervalIndex` ? 
Our public API is already huge, not sure this is needed to add even more. We have other examples where Index subclasses have specific type-related methods that are not available for other Index classes.

File: doc/source/whatsnew/v0.23.0.txt [hunk 13]
```diff
@@ -323,6 +323,7 @@ Other Enhancements
 
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
+- New :meth:`IntervalIndex.get_loc_exact` has been added to find exact Interval matches only (:issue:`19349`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: prob needs to be added to api.rst

File: pandas/core/indexes/base.py [hunk 14]
```diff
@@ -2536,6 +2536,21 @@ def get_loc(self, key, method=None, tolerance=None):
             raise KeyError(key)
         return loc
 
+    def get_loc_exact(self, key, method=None):
+        """Get integer location, slice or boolean mask for exact
+        matches only.
+
```
Comment (jreback) [inline pandas/core/indexes/base.py]: yeah could remove this

File: asv_bench/benchmarks/indexing.py [hunk 15]
```diff
@@ -229,6 +229,10 @@ def time_getitem_list(self, monotonic):
     def time_loc_list(self, monotonic):
         monotonic.loc[80000:]
 
+    def time_get_loc_exact(self, monotonic):
+        interval = monotonic.index[80000]
+        monotonic.index.get_loc_exact(interval)
```
Comment (jreback) [inline asv_bench/benchmarks/indexing.py]: should only run on II

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jschendel** [comment]
I was just going over some of the tests in `test_interval_new.py`, and noticed the following comment in `test_get_loc_interval`, which indicates that the desired future behavior for `get_loc` is to only return exact matches:

https://github.com/pandas-dev/pandas/blob/324379ce75269aa6bced90ecf3edb692539a2742/pandas/tests/indexes/interval/test_interval_new.py#L24-L31

Is this really the desired future behavior?  If so, should we try putting the `get_loc_exact` behavior in this PR directly into the `Interval` cases for `get_loc`?  Couldn't find any comments specifically about this after doing a quick search through the issue/pr that added the new test.

**Comment by jorisvandenbossche** [comment]
@jschendel Yes, that was also my first thought when I saw this PR, and wanted to say: let's change `

**Comment by jorisvandenbossche** [comment]
@jschendel Yes, that was also my first thought when I saw this PR, and wanted to say: let's change `get_loc` to be exact as we discussed that we want it this way. 
However, the comment you cite above is when it is passed an *interval*, then it is required to be exact (currently you can also pass a 'sub-interval'). But for non-interval scalars it still works as well, so eg `df.loc[1]` will still work to select all rows with intervalindex where the interval contains 1. So for scalars it does not need to be exact (at least, that is what we discussed then).

**Comment by jorisvandenbossche** [comment]
So it is the test below the one you quoted (`test_get_loc_scalar`): https://github.com/pandas-dev/pandas/blob/324379ce75269aa6bced90ecf3edb692539a2742/pandas/tests/indexes/interval/test_interval_new.py#L57-L58

**Comment by jschendel** [comment]
@jorisvandenbossche : Yes, maybe it wasn't clear from my comment, but I was suggesting just modifying the `Interval` specific behavior of `get_loc` to force exactness.  Since we're already writing the logic to do this, I thought it might not be that much additional work to just incorporate it in `get_loc` as well, which could lead to some simplifications for `get_loc_exact`.  After looking over the code a bit more though, it might not be as straightforward as I initially thought.

**Comment by jorisvandenbossche** [comment]
> modifying the Interval specific behavior of get_loc to force exactness.

Ah, yes, I missed that. But that's a good idea
